src/backend/drm/renderer/mod.rs

```diff
@@ -0,0 +1,2553 @@
+                        dst: config.dst,
+                        transform: config.transform,
+                        damage_clips: config.damage_clips.as_ref().map(|d| d.blob()),
+                        fb: *config.buffer.as_ref(),
+                    }),
+                }),
+            event,
+        )
+    }
+}
+
+/// Possible buffers to export as a framebuffer using [`ExportFramebuffer`]
+#[derive(Debug)]
+pub enum ExportBuffer<'a, B: Buffer> {
+    /// A wayland buffer
+    Wayland(&'a WlBuffer),
+    /// A [`Allocator`] buffer
+    Allocator(&'a B),
+}
+
+impl<'a, B: Buffer> From<&'a UnderlyingStorage> for ExportBuffer<'a, B> {
+    fn from(storage: &'a UnderlyingStorage) -> Self {
+        match storage {
+            UnderlyingStorage::Wayland(buffer) => Self::Wayland(buffer),
+        }
+    }
+}
+
+/// Export a [`ExportBuffer`] as a framebuffer
+pub trait ExportFramebuffer<B: Buffer>
+where
+    B: Buffer,
+{
+    /// Type of the framebuffer
+    type Framebuffer: AsRef<framebuffer::Handle>;
+
+    /// Type of the error
+    type Error: std::error::Error;
+
+    /// Add a framebuffer for the specified buffer
+    fn add_framebuffer(
+        &self,
+        drm: &DrmDeviceFd,
+        buffer: ExportBuffer<'_, B>,
+    ) -> Result<Option<Self::Framebuffer>, Self::Error>;
+}
+
+impl<F, B> ExportFramebuffer<B> for Arc<Mutex<F>>
+where
+    F: ExportFramebuffer<B>,
+    B: Buffer,
+{
+    type Framebuffer = <F as ExportFramebuffer<B>>::Framebuffer;
+    type Error = <F as ExportFramebuffer<B>>::Error;
+
+    fn add_framebuffer(
+        &self,
+        drm: &DrmDeviceFd,
+        buffer: ExportBuffer<'_, B>,
+    ) -> Result<Option<Self::Framebuffer>, Self::Error> {
+        let guard = self.lock().unwrap();
+        guard.add_framebuffer(drm, buffer)
+    }
+}
+
+impl<F, B> ExportFramebuffer<B> for Rc<RefCell<F>>
+where
+    F: ExportFramebuffer<B>,
+    B: Buffer,
+{
+    type Framebuffer = <F as ExportFramebuffer<B>>::Framebuffer;
+    type Error = <F as ExportFramebuffer<B>>::Error;
+
+    fn add_framebuffer(
+        &self,
+        drm: &DrmDeviceFd,
+        buffer: ExportBuffer<'_, B>,
+    ) -> Result<Option<Self::Framebuffer>, Self::Error> {
+        self.borrow().add_framebuffer(drm, buffer)
+    }
+}
+
+/// Specialized renderer utilizing drm planes with support for direct scan-out
+#[derive(Debug)]
+pub struct DrmRenderer<A, F, U>
+where
+    A: Allocator,
+    F: ExportFramebuffer<A::Buffer>,
+    <F as ExportFramebuffer<A::Buffer>>::Framebuffer: std::fmt::Debug + 'static,
+{
+    output: Output,
+    surface: Arc<DrmSurface>,
+    planes: Planes,
+    damage_tracked_renderer: DamageTrackedRenderer,
+
+    framebuffer_exporter: F,
+
+    current_frame: FrameState<DrmScanoutBuffer<A::Buffer, <F as ExportFramebuffer<A::Buffer>>::Framebuffer>>,
+    #[allow(clippy::type_complexity)]
+    pending_frame: Option<(
+        FrameState<DrmScanoutBuffer<A::Buffer, <F as ExportFramebuffer<A::Buffer>>::Framebuffer>>,
```

Comment:
I'd actually agree with clippy here. Maybe a Frame type internally used could be made

Reply:
Yeah, I will try to move that to a type def.

Reply:
@i509VCB I added some internal type defs and removed all `[allow(clippy::type_complexity)]`